updater/src/updater.rs

```diff
@@ -255,47 +380,151 @@ impl Updater {
-			let auto = {
-				let mut s = self.state.lock();
-				let fetched = s.fetching.take().unwrap();
-				let dest = self.updates_path(&Self::update_file_name(&fetched.version));
-				if !dest.exists() {
-					let b = match result {
-						Ok(b) => {
-							s.backoff = None;
-							b
-						},
-						Err(e) => {
-							let mut n = s.backoff.map(|b| b.0 + 1).unwrap_or(1);
-							s.backoff = Some((n, Instant::now() + Duration::from_secs(2usize.pow(n) as u64)));
-
-							return Err((format!("Unable to fetch update ({}): {:?}", fetched.version, e), false));
-						},
-					};
-
-					info!(target: "updater", "Fetched latest version ({}) OK to {}", fetched.version, b.display());
-					fs::create_dir_all(dest.parent().expect("at least one thing pushed; qed")).map_err(|e| (format!("Unable to create updates path: {:?}", e), true))?;
-					fs::copy(&b, &dest).map_err(|e| (format!("Unable to copy update: {:?}", e), true))?;
-					restrict_permissions_owner(&dest, false, true).map_err(|e| (format!("Unable to update permissions: {}", e), true))?;
-					info!(target: "updater", "Installed updated binary to {}", dest.display());
-				}
-				let auto = match self.update_policy.filter {
-					UpdateFilter::All => true,
-					UpdateFilter::Critical if fetched.is_critical /* TODO: or is on a bad fork */ => true,
-					_ => false,
+	fn updater_step(&self) {
+		let current_block_number = self.client.upgrade().map_or(0, |c| c.block_number(BlockId::Latest).unwrap_or(0));
+
+		let mut state = self.state.lock();
+
+		if let Some(latest) = state.latest.clone() {
+			let fetch = |binary| {
+				info!(target: "updater", "Attempting to get parity binary {}", binary);
+				let weak_self = self.weak_self.lock().clone();
+				let latest = latest.clone();
+				let on_fetch = move |res: Result<PathBuf, fetch::Error>| {
+					if let Some(this) = weak_self.upgrade() {
+						let mut state = this.state.lock();
+
+						// Check if the latest release and updater status hasn't changed
+						if state.latest.as_ref() == Some(&latest) {
+							if let UpdaterStatus::Fetching { ref release, binary, retries } = state.status.clone() {
+								match res {
+									// We've successfully fetched the binary
+									Ok(path) => {
+										let setup = |path: &Path| -> Result<(), String> {
+											let dest = this.updates_path(&Self::update_file_name(&release.version));
+											if !dest.exists() {
+												info!(target: "updater", "Fetched latest version ({}) OK to {}", release.version, path.display());
+												fs::create_dir_all(dest.parent().expect("at least one thing pushed; qed")).map_err(|e| format!("Unable to create updates path: {:?}", e))?;
+												fs::copy(path, &dest).map_err(|e| format!("Unable to copy update: {:?}", e))?;
+												restrict_permissions_owner(&dest, false, true).map_err(|e| format!("Unable to update permissions: {}", e))?;
+												info!(target: "updater", "Copied updated binary to {}", dest.display());
+											}
```

Comment:
> The answer to that is that if you need more than 3 levels of indentation, you're screwed anyway, and should fix your program.

– Linus Torvalds

No but seriously though, is there a way to break this up somehow?

Reply:
please exit fast on if let statements, cause it's difficult to read it :)

Reply:
Agreed, I'll clean it up.